src/main/java/edu/harvard/iq/dataverse/DOIDataCiteRegisterService.java

```diff
@@ -406,10 +450,74 @@ public String generateXML() {
             }
             contributorsElement.append("</contributor>");
         }
+        
+        String relIdentifiers = generateRelatedIdentifiers (dvObject);
+        
+        xmlMetadata = xmlMetadata.replace("${relatedIdentifiers}", relIdentifiers);
+
         xmlMetadata = xmlMetadata.replace("{$contributors}", contributorsElement.toString());
         return xmlMetadata;
     }
 
+    private String generateRelatedIdentifiers(DvObject dvObject) {
+
+        StringBuilder sb = new StringBuilder();
+        if (dvObject.isInstanceofDataset()) {
+            Dataset dataset = (Dataset) dvObject;
+            if (!dataset.getFiles().isEmpty() && !(dataset.getFiles().get(0).getIdentifier() == null)) {
+
+                datafileIdentifiers = new ArrayList<>();
+                for (DataFile dataFile : dataset.getFiles()) {
+                    String add = "";
+                    sb.append("");
+                    if (!dataFile.getGlobalId().asString().isEmpty()) {
+                        add = "<relatedIdentifier relatedIdentifierType=\"DOI\" relationType=\"HasPart\""
+                                + ">" + dataFile.getGlobalId() + "</relatedIdentifier>";
+                        if (sb.toString().isEmpty()) {
+                            sb.append("<relatedIdentifiers>");
+                        }
+                        sb.append(add);
+                    }
+                }
+
+                if (!sb.toString().isEmpty()) {
+                    sb.append("</relatedIdentifiers>");
+                }
+            }
+        } else if (dvObject.isInstanceofDataFile()){
```

Comment:
Is there ever a case where we have a data file but for some reason it doesn't know its owner?

Reply:
I'm pretty sure the answer to that is no. Data files will have an owner if they are in a dataset.